### PR TITLE
Use parity-common to make substrate generic over hasher and trie encoding codec

### PR DESCRIPTION
**Pull request by dvdplm**
As part of the effort to extract and use common code between `parity-ethereum` and `polkadot`, this PR introduces the crates in `parity-common` which in turn means switch to using generics for the hash type and node encoder and add concrete impls (`BlakeHasher` and `BlakeRlpCodec`).
Part of the code is completely generic, e.g. `state-machine`, whereas other parts use `BlakeHasher` and `BlakeRlpCodec` to build concrete types. The exact layering of which parts are generic and which aren't is not very well thought out and feedback on how to improve that is most welcome.

### CODE REVIEW

File: substrate/state-machine/src/backend.rs [hunk 1]
```diff
@@ -66,21 +74,25 @@ impl error::Error for Void {
 /// In-memory backend. Fully recomputes tries on each commit but useful for
 /// tests.
 #[derive(Clone, PartialEq, Eq)]
-pub struct InMemory {
+pub struct InMemory<H, C> {
 	inner: Arc<HashMap<Vec<u8>, Vec<u8>>>,
+	_hasher: PhantomData<H>,
```
Comment (gnunicorn) [inline substrate/state-machine/src/backend.rs]: Curious: why did you opt for an underscore here?

File: substrate/state-machine/src/ext.rs [hunk 2]
```diff
@@ -123,7 +136,7 @@ impl<'a, B: 'a> Externalities for Ext<'a, B>
 		42
 	}
 
-	fn storage_root(&mut self) -> [u8; 32] {
+	fn storage_root(&mut self) -> H::Out where H::Out: Ord + Encodable {
```
Comment (gnunicorn) [inline substrate/state-machine/src/ext.rs]: I know this line isn't super long, however I think it reduces readability complexity if we had the `where` in the second line

File: substrate/state-machine/src/testing.rs [hunk 3]
```diff
@@ -17,48 +17,64 @@
...
-/// Simple HashMap based Externalities impl.
-pub type TestExternalities = HashMap<Vec<u8>, Vec<u8>>;
+/// Simple HashMap-based Externalities impl.
+pub struct TestExternalities<H> {
+	inner: HashMap<Vec<u8>, Vec<u8>>,
+	hasher: PhantomData<H>,
+}
+
+impl<H: Hasher> TestExternalities<H> {
+	#[cfg(test)]
```
Comment (gnunicorn) [inline substrate/state-machine/src/testing.rs]: is this `#[cfg]` necessary?

File: substrate/state-machine/src/proving_backend.rs [hunk 4]
```diff
@@ -50,25 +50,30 @@ impl ProvingBackend {
...
+		let map_e = |e| format!("Trie lookup error: {}", e);
 
 		let mut proof_recorder = self.proof_recorder.try_borrow_mut()
 			.expect("only fails when already borrowed; storage() is non-reentrant; qed");
-		TrieDB::new(&eph, &self.backend.root()).map_err(map_e)?
+		TrieDB::<H, C>::new(&eph, &self.backend.root()).map_err(map_e)?
 			.get_with(key, &mut *proof_recorder).map(|x| x.map(|val| val.to_vec())).map_err(map_e)
 	}
 
+	// TODO: "error[E0283]: type annotations required: cannot resolve `_: patricia_trie::NodeCodec<H>`"/"note: required because of the requirements on the impl of `backend::Backend<H, _>` for `trie_backend::TrieBackend<H>`")
```
Comment (gnunicorn) [inline substrate/state-machine/src/proving_backend.rs]: Also comments shouldn't be longer than 100chars, please wrap :) .

File: substrate/keystore/Cargo.toml [hunk 5]
```diff
@@ -4,7 +4,7 @@ version = "0.1.0"
 authors = ["Parity Technologies <admin@parity.io>"]
 
 [dependencies]
-ethcore-crypto = { git = "https://github.com/paritytech/parity.git", default_features = false }
+parity-crypto = { git = "https://github.com/paritytech/parity-common.git", default_features = false }
```
Comment (gnunicorn) [inline substrate/keystore/Cargo.toml]: You missed renaming an import, [see travis test](https://travis-ci.org/paritytech/polkadot/jobs/402680325#L1050).

File: substrate/state-machine/src/backend.rs (hunk 1 above)
Comment (dvdplm) [reply substrate/state-machine/src/backend.rs]: No real good reason. I keep going back and forth on what how to name my phantoms, so I end up being inconsistent. Sometimes I like `marker`, sometimes I name it for what it contains, sometimes with a `_` prefix to signal to the reader "never mind this, it's there to make the code compile and is not relevant".

I'd like to have a rule to follow here. What do you use?

File: substrate/state-machine/src/testing.rs (hunk 3 above)
Comment (dvdplm) [reply substrate/state-machine/src/testing.rs]: I think it's wrong. `TestExternalities` is exported (and re-exported again in `runtime-io`) so I think it needs to be `pub`. Good catch.

Comment (dvdplm) [reply substrate/state-machine/src/testing.rs]: …and this change will break things in other crates. I don't think this PR can be merged as-is.

File: substrate/keystore/Cargo.toml (hunk 5 above)
Comment (dvdplm) [reply substrate/keystore/Cargo.toml]: fixed, thanks!

File: substrate/state-machine/src/backend.rs (hunk 1 above)
Comment (gnunicorn) [reply substrate/state-machine/src/backend.rs]: [Clippy states that underscore-prefixed names](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#used_underscore_binding) are expected to be in there for compile-passing-purposes only and aren't actually used. Which is the case here, and combined with a proper descriptive name, I think this is totally the appropriate way to do it, @dvdplm  :) .

Comment (dvdplm) [reply substrate/state-machine/src/backend.rs]: "compile-passing-purposes" – rofl

**Comment by arkpar** [comment]
Closes #292

**Comment by rphmeier** [comment]
@dvdplm yes, we should get a `BlakeTwo256` and maybe a `substrate-codec` RLP codec. Or we could pull in the `RlpNodeCodec`

**Comment by gavofyork** [comment]
What's going on with this PR? Ready for merge?

**Comment by dvdplm** [comment]
> What's going on with this PR? Ready for merge?

No.
The state-machine code changed here is fine but if merged it will break things all over. After talking this over with @rphmeier we concluded that we might as well bite the bullet and make substrate generic over the hashing algorithm and codec all at once.

File: substrate/client/src/block_builder.rs [hunk 6]
```diff
@@ -23,12 +23,20 @@ use runtime_primitives::traits::{Header as HeaderT, Hash, Block as BlockT, One,
...
 /// Utility for building new (valid) blocks from a stream of extrinsics.
-pub struct BlockBuilder<B, E, Block> where
-	B: backend::Backend<Block>,
-	E: CallExecutor<Block> + Clone,
+pub struct BlockBuilder<B, E, Block, H, C>
+where
+	B: backend::Backend<Block, H, C>,
+	E: CallExecutor<Block, H, C> + Clone,
 	Block: BlockT,
+	H: Hasher,
```
Comment (gavofyork) [inline substrate/client/src/block_builder.rs]: reason for not integrating `H` and `C` into `BlockT`?

File: substrate/primitives/src/rlp_codec.rs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+		if r.is_data() && r.size() == BlakeHasher::LENGTH {
+			Some(r.as_val().expect("Hash is the correct size; qed"))
+		} else {
+			None
+		}
+	}
+	fn is_empty_node(data: &[u8]) -> bool {
+		Rlp::new(data).is_empty()
+	}
+    fn empty_node() -> ElasticArray1024<u8> {
```
Comment (gavofyork) [inline substrate/primitives/src/rlp_codec.rs]: get them spaces out of here!!!!

File: substrate/primitives/src/rlp_codec.rs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+                let bytes = &AsRef::<[u8]>::as_ref(&inline_data)[..len];
+                stream.append_raw(bytes, 1)
+            },
+        };
+        stream.drain()
+	}
+
+	fn branch_node<I>(children: I, value: Option<ElasticArray128<u8>>) -> ElasticArray1024<u8>
+	where I: IntoIterator<Item=Option<ChildReference<<BlakeHasher as Hasher>::Out>>>
+    {
```
Comment (gavofyork) [inline substrate/primitives/src/rlp_codec.rs]: gah the indenting is borked. spaces!

**Comment by gavofyork** [comment]
two concerns:
1. same as @svyatonik - why do I see `BlakeHasher` so frequently in substrate - isn't the point to keep it generic? it looks very much like most of the runtime and much of the backend can only be used with BlakeHasher, drastically reducing this PR's benefit.
2. there's a lot of extra code noise with these changes. i'm thinking of (quite numerous) instances of:
```
- ) -> Result<(Vec<u8>, B::Transaction), Box<Error>> {
+ ) -> Result<(Vec<u8>, B::Transaction), Box<Error>>
+ where
+ 	H: Hasher,
+ 	C: NodeCodec<H>,
+ 	Exec: CodeExecutor<H>,
+ 	B: Backend<H, C>,
+ 	H::Out: Ord + Encodable,
+ 	Handler: FnOnce(Result<Vec<u8>, Exec::Error>, Result<Vec<u8>, Exec::Error>) -> Result<Vec<u8>, Exec::Error>
+ {
```

**Comment by dvdplm** [comment]
> I thought that the point of the PR is to avoid requirements for concrete implementations in substrate.

Up to a point yes, the goal is to make things generic but the main motivating force was to share code with `parity-ethereum`. There is something of a "hard limit" as to how generic substrate can be, though. I found it quite tricky to make things generic beyond this [here](https://github.com/paritytech/polkadot/pull/297/files#diff-8676d22639686f518b914ccd4dcb4b7eR39):

```
environmental!(ext: trait Externalities<BlakeHasher>);
```

That line can not not use concrete types, and this somewhat limits the extent to which the rest can be generic. I am sure I've missed places that could be generic and instead went for concrete types (do holler if you spot some!).

The result is "a lot of BlakeHasher" all over. I think this can be alleviated in many places with some convenience types but I'd like to address that in a separate PR. I think it is easier to get good review feedback with all the nastiness laid bare and then sweep stuff under the rug in a a phase 2. If you disagree I'm happy to do it right away.

**Comment by gavofyork** [comment]
ok - make it an issue and i guess we can revisit.

File: substrate/client/src/block_builder.rs (hunk 6 above)
Comment (dvdplm) [reply substrate/client/src/block_builder.rs]: Not sure I understand what you mean here. You mean something like this?

```
pub trait Block: Clone + Send + Sync + Codec + Eq + MaybeSerializeDebug + 'static {
	type Extrinsic: Member + Codec;
	type Header: Header<Hash=Self::Hash>;
	type Hash: Member + ::rstd::hash::Hash + Copy + MaybeDisplay + Default + SimpleBitOps + Codec + AsRef<[u8]>;
	type Hasher: Hasher<Out=<Self::Hash>::Output>;
	type NodeCodec: NodeCodec<Self::Hasher>;
…
```

There is some significant overlap between the `hashdb::Hasher` and substrate own `Hash` trait and when talking this over with @rphmeier his idea was to get this in first and refine/unify further down the line. Not sure if `Block` is the proper unification point but might very well be?

**Pull request merged**